### ISSUE TITLE
JUnitReportReporter doesn't report configuration errors

### ISSUE THREAD

**Issue by martoe** [cbeust/testng#1149]
(Follow-up to issue #1078)

The XML report produced by JUnitReportReporter is incomplete:
If a configuration method fails, the dependent tests are now reported as "skipped", but the original error is not shown.
I guess the correct behaviour would be to additionally include the failed configuration method in the report (like the HTML reporter does)

**Comment by juherr** [comment]
I don't know what to think about this issue because it looks to be the expected behavior since https://github.com/cbeust/testng/commit/3c331730b067c4367825ba373dbcf0be02df6fe5

What do you think @cbeust ?

**Comment by martoe** [comment]
Take this test class as illustration:
```
@Test
public class ConfigFailTest {
  
  private static int count = 1;
  
  @BeforeMethod // unstable config method, fails on second test
  public void setup() {
    if (count++ == 2) {
      throw new RuntimeException();
    }
  }
  public void test1() {}
  public void test2() {}
  public void test3() {}
}
```
The TestNG HTML report shows **one passed method** (test1), **two skipped methods** (test2+test3) and **one failed config-method**. This is perfectly reasonable.
The JUnit report XML on the other hand contains **one passed method** (test1) and **two failed methods** (test2+test3), which looks wrong to me. On larger testsuites, it is really hard to find out which test actually went wrong in such a szenario.

**Comment by juherr** [comment]
Yes, it looks logic for me too, but I'm just highlighting the fact that is was intentional.

I saw you already fixed the problem, then you can already propose it and will see @cbeust 's decision there.

**Comment by cbeust** [comment]
@martoe JUnit doesn't have the concept of a "skipped" test, all it knows is "passed" or "failed", hence the report you're seeing.

This was one of my original motivation to start TestNG in 2003 :-)

**Comment by martoe** [comment]
@cbeust JUnit might not support skipped tests, but at least the JUnit XML report format used by Jenkins-CI supports it - and the current JUnitReportReporter already implements this (line 118), but the implementation is not correct.

To reuse the previous example, the correct (Jenkins-complient) output should be:
```
<?xml version="1.0" encoding="UTF-8"?>
<!-- Generated by org.testng.reporters.JUnitReportReporter -->
<testsuite skipped="2" hostname="latitude" name="ConfigFailTest" tests="4" failures="0" timestamp="13 Sep 2016 18:34:44 GMT" time="0.017" errors="1">
  <testcase name="test3" time="0.013" classname="ConfigFailTest">
    <skipped/>
  </testcase> <!-- test3 -->
  <testcase name="test1" time="0.002" classname="ConfigFailTest"/>
  <testcase name="setup" time="0.000" classname="ConfigFailTest">
    <error type="java.lang.RuntimeException">
      <![CDATA[java.lang.RuntimeException
	at ConfigFailTest.setup(ConfigFailTest.java:11)
[...]
]]>
    </error>
  </testcase> <!-- setup -->
  <testcase name="test2" time="0.002" classname="ConfigFailTest">
    <skipped/>
  </testcase> <!-- test2 -->
</testsuite> <!-- ConfigFailTest -->
```

**Comment by cbeust** [comment]
Ah I didn't realize that. Do you know how "not correct" the implementation is? Would you be willing to send a PR?

**Comment by juherr** [comment]
@cbeust You can check his commit here: https://github.com/martoe/testng/commit/2d0fa835e9a585acba65c9ec5944db6a6f0c6f4e which is more/less a revert of https://github.com/cbeust/testng/commit/3c331730b067c4367825ba373dbcf0be02df6fe5

**Comment by martoe** [comment]
https://github.com/cbeust/testng/pull/1152 - but no unittest yet :(

**Comment by juherr** [comment]
Closed by #1152
